### PR TITLE
solution for pair programming

### PR DESCRIPTION
**Pull request by devo1ution**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by devo1ution** [comment]
hi guys I would appreciate if anyone could leave some feedback on the partial solution I've written thanks :)

File: solutions/gold/usaco-1234.mdx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+2. As long as $b_j$ isn't $*0$, $\boxed{|B| = |I(i, j - 1)|}$
+3. $I(i, j) = A \cup B$
+
+Take some time to convince yourself of these, especially points 1 and 2!
+
+At this point, you may be tempted to conclude that:
+
+$|I(i, j)| = |A \cup B| = |A| + |B| = |I(i - 1, j)| + |I(i, j - 1)|$.
+
+Don't make this mistake! Remember the **Principle of Inclusion-Exclusion**:
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1234.mdx]: wikipedia link?

File: solutions/gold/usaco-1234.mdx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+<Info title="Notation">
+In the below solution:
+- $|S|$ denotes the number of elements in a set $S$
+- $S \cup T$ denotes the union of sets $S$ and $T$ (elements in either $S$ or $T$)
+- $S \cap T$ denotes the intersection of sets $S$ and $T$ (elements in both $S$ and $T$)
+- $\text{set of expressions} \rightarrow \text{instruction}$ denotes applying that instruction to each expression in the set
+</Info>
+So, how do we calculate $|I(i, j)|$?
+
+When either $i$ or $j$ are 0, the interleaving is uniquely determined and thus $\boxed{|I(i, 0)| = |I(0, i)| = 1}$ for all $i \leq N$.
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1234.mdx:34]: why's this?

File: solutions/gold/usaco-1234.mdx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+Great, but what if $i$ and $j$ are both nonzero? Evidently, any interleaving of Bessie's first $i$ instructions and Elsie's first $j$ instructions must either end with instruction $a_i$ or instruction $b_j$. We can thus rewrite the expressions that end with $a_i$ as $I(i - 1, j) \rightarrow a_i$ (let's denote this set as $A$) and the ones that end with $b_j$ as $I(i, j - 1) \rightarrow b_j$ (let's denote this set as $B$). We also have the following:
+
+1. As long as $a_i$ isn't $*0$, $\boxed{|A| = |I(i - 1, j)|}$
+2. As long as $b_j$ isn't $*0$, $\boxed{|B| = |I(i, j - 1)|}$
+3. $I(i, j) = A \cup B$
+
+Take some time to convince yourself of these, especially points 1 and 2!
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1234.mdx:43]: is this one of those "exercise is left to the reader" things

**Review by SansPapyrus683 — COMMENTED** [review]
very bing chilling editorial

File: solutions/gold/usaco-1234.mdx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+$$
+|I(i, j)| = |A \cup B| = |A| + |B| = |I(i - 1, j)| + |I(i, j - 1)|
+$$
+
+Don't make this mistake!
+Remember the **Principle of Inclusion-Exclusion**:
+
+$\boxed{|S \cup T| = |S| + |T| - |S \cap T|}$
+
```
Comment (devo1ution) [inline solutions/gold/usaco-1234.mdx:55]: not sure how to format this section nicely

**Comment by devo1ution** [comment]
thanks for the feedback hopefully i finish this sol by the end of the week 🙏🏻

**Comment by SansPapyrus683** [comment]
ayo any progress on this pr?

File: solutions/gold/usaco-1234.mdx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+Take some time to convince yourself of these, especially points 1 and 2!
+
+<Info title="Not Convinced?">
+
+Use the fact that if we apply the same, non-$*0$ operation to two distinct expressions, the results will still be distinct.
+
+</Info>
+
+At this point, you may be tempted to conclude that
+
```
Comment (devo1ution) [inline solutions/gold/usaco-1234.mdx:53]: how do i get rid of 3.1/3.2 numbering here

File: solutions/gold/usaco-1234.mdx (hunk 3 above)
Comment (devo1ution) [reply solutions/gold/usaco-1234.mdx:44]: yep

File: solutions/gold/usaco-1234.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/gold/usaco-1234.mdx:53]: wdym?

File: solutions/gold/usaco-1234.mdx [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+const int N = 2e3 + 1;
+const int MOD = 1e9 + 7;
+
+char a[N], b[N];
+
+// zero[i][j] -> whether 0 is in I(i, j)
+bool zero[N][N];
+
+// number of nonzero expressions in I(i, j)
+ll dp[N][N];
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1234.mdx:119]: please use vectors instead

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/usaco-1234.mdx (hunk 5 above)
Comment (devo1ution) [reply solutions/gold/usaco-1234.mdx:53]: if you look at the preview where it should say 
1. $a_i$ and $b_j$ are commutative
2. the expression can be written as...
it says
3.1 $a_i$...
3.2 the expression....
instead

File: solutions/gold/usaco-1234.mdx [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+const int MOD = 1e9 + 7;
+
+vector<char> a, b;
+
+// zero[i][j] -> whether 0 is in I(i, j)
+vector<vector<bool>> zero;
+
+// dp[i][j] -> number of nonzero expressions in I(i, j)
+vector<vector<ll>> dp;
+bool digit(char c) { return c >= '1' && c <= '9'; }
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1234.mdx:120]: actually i think there's a built-in func for this
https://en.cppreference.com/w/cpp/string/byte/isdigit

File: solutions/gold/usaco-1234.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/gold/usaco-1234.mdx:53]: might be a bug in mdx tbh, it doesn't matter too much

File: solutions/gold/usaco-1234.mdx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+const int N = 2e3 + 1;
+const int MOD = 1e9 + 7;
+
+vector<char> a, b;
+
+// zero[i][j] -> whether 0 is in I(i, j)
+vector<vector<bool>> zero;
+
+// dp[i][j] -> number of nonzero expressions in I(i, j)
+vector<vector<ll>> dp;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1234.mdx:119]: these have no need to be global

File: solutions/gold/usaco-1234.mdx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+			dp[i][0] = !zero[i][0];
+		}
+
+		for (int i = 1; i <= n; i++) {
+			for (int j = 1; j <= n; j++) {
+				dp[i][j] = 0;
+				if (a[i] != '0') ad(dp[i][j], dp[i - 1][j]);
+				if (a[i] == '+') ad(dp[i][j], zero[i - 1][j]);
+				if (b[j] != '0') ad(dp[i][j], dp[i][j - 1]);
+				if (b[j] == '+') ad(dp[i][j], zero[i][j - 1]);
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1234.mdx:168]: braces

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by devo1ution** [comment]
@SansPapyrus683 could you review again?

**Review by SansPapyrus683 — APPROVED** [review]
we makin' it outta the barn with this one

@ryanchou-dev

**Review by ryanchou-dev — COMMENTED** [review]
fire explanation

**Pull request merged**